### PR TITLE
[dynamo] Stop writing a source onto the interned ConstantVariable(None)

### PR DESCRIPTION
**Pull request by bobrenjc93**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.14.0) (oldest at bottom):
* #195070
* #195014
* #194966
* #194952
* #194951
* #194950
* #194949
* #194948
* #194677
* #194629
* #194622
* #194619
* #194549
* #194542
* #194541
* #194534
* #194488
* #194479
* #194455
* #194454
* #194453
* #194452
* #194451
* #194450
* #194449
* #194428
* #194427
* #194426
* #194409
* #194394
* #194393
* #194390
* #194389
* #194388
* #194387
* #194386
* #194385
* #194384
* #194383
* #194320
* #194319
* #194318
* #194317
* #194316
* #194302
* #194301
* #194300
* #194286
* #194285
* #194284
* #194283
* #194282
* #194250
* #194249
* #194248
* #194158
* #192379
* #193977
* #193976
* #193697
* #192374
* #192866
* __->__ #193725

First of six fixes this stack needs before torch.compiler.precompile at the
top of it can capture anything.

Background, because the bug is only obvious once two facts sit next to each
other. Dynamo never handles real values while tracing; every value is wrapped
in a VariableTracker, and `ConstantVariable` is the wrapper for the ones it can
treat as compile-time constants. `ConstantVariable.create(None)` therefore does
not make a None -- it asks for the tracker that DESCRIBES None. And because
every None is indistinguishable, `create` is a cache: with no kwargs it returns
`CONSTANT_VARIABLE_NONE`, one object allocated at import and handed to every
caller in the process. The second fact is that a tracker's `source` records how
to rebuild the value, in bytecode, from the frame currently being compiled --
`LOAD_FAST x; LOAD_ATTR grad`. It is meaningful only inside that frame.

Reading a tensor attribute labels the result with an AttrSource so downstream
code can reconstruct it. `getattro_impl` did that by assignment, which is safe
only when the result is a tracker it just built. For `.grad` and
`requires_grad` it is not: the value is the shared singleton, so the assignment
stamps a name from THIS frame onto the object every later compile in the
process will also be handed. The next unrelated function to touch a None gets a
tracker claiming it can be rebuilt as `x.grad`, tries it, and dies in
`create_load` with "<name> missing" -- in code with no connection to the
culprit. The cache is fine; mutating a cached object is not, for the same
reason you do not mutate a default argument.

Clone before labelling, but only for constants. A blanket clone breaks `.data`:
the tracker it hands back is freshly created, and `VariableTracker.__init__`
rejects a source on one of those ("source must be None for
ValueMutationNew/AttributeMutationNew"), so only the in-place write gets
through. The guard above the branch already restricts constants to
grad/requires_grad, so the clone arm is exactly the singleton case.

Why the top of the stack needs it: a precompile training capture pins
`trace_autograd_ops` and seeds a zero `.grad` on every autograd target before
re-capturing, which is exactly the shape that poisons the singleton. With this
hunk reverted, 32 tests in test/test_precompile.py fail.

Test Plan:

The regression test turns on `trace_autograd_ops`, compiles a step that nulls
`.grad` before calling `backward()`, then compiles an unrelated function. The
second compile raises without the fix.

```
python test/dynamo/test_repros.py
python test/dynamo/test_misc.py
```

Authored with an AI assistant.

Differential Revision: [D116222525](https://our.internmc.facebook.com/intern/diff/D116222525)

### CODE REVIEW

**Comment by bobrenjc93** [comment]
@bobrenjc93 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D116222525).

**Comment by bobrenjc93** [comment]
@bobrenjc93 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D116222525).

**Comment by bobrenjc93** [comment]
@bobrenjc93 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D116222525).

**Comment by bobrenjc93** [comment]
Closing: landed on main via #194749.
